rust/bindgen.bzl

```diff
@@ -0,0 +1,156 @@
+# limitations under the License.
+
+def _rust_bindgen(ctx):
+    rust_toolchain = ctx.toolchains["@io_bazel_rules_rust//rust:toolchain"]
+
+    bindgen = ctx.executable.bindgen
+    rustfmt = rust_toolchain.rustfmt if hasattr(rust_toolchain, 'rustfmt') else None
+    clang = ctx.executable.clang
+    libclang = ctx.attr.libclang
+    libstdcxx = ctx.attr.libstdcxx
+    cc_toolchain = ctx.attr.cc_toolchain
+    cc_fragment = ctx.fragments.cpp
+
+    # nb. We can't grab the cc_library`s direct headers, so a header must be provided.
+    cc_lib = ctx.attr.cc_lib
+    if not hasattr(cc_lib, "cc"):
+        fail("{} is not a cc_library".format(cc_lib))
+    header = ctx.file.header
+    if header not in cc_lib.cc.transitive_headers:
+        fail("Header {} is not in {}'s transitive closure of headers.".format(ctx.attr.header, cc_lib))
+
+    # rustfmt is not in the usual place, so bindgen would fail to find it
+    bindgen_args = ["--no-rustfmt-bindings"]
+    clang_args = []
+
+    output = ctx.outputs.out
+
+    clang_libs = depset(libclang.cc.libs + libstdcxx.cc.libs)
+    include_directories = depset(
+        cc_fragment.built_in_include_directories + [f.dirname for f in cc_lib.cc.transitive_headers]
+    )
+
+    if rustfmt:
+        unformatted = ctx.actions.declare_file(output.basename + ".unformatted")
+    else:
+        unformatted = output
+
+    args = ctx.actions.args()
+    args.add_all(bindgen_args)
+    args.add(header.path)
+    args.add("--output", unformatted.path)
+    args.add("--")
+    args.add_all(include_directories, before_each="-I")
+    args.add_all(clang_args)
+    ctx.actions.run(
+        inputs=depset(
+            [header],
+            # Standard library headers aren't in the transitive_headers, so we find them in the cc_toolchain
+            transitive=[cc_lib.cc.transitive_headers, clang_libs, cc_toolchain.files],
+        ),
+        outputs=[unformatted],
+        mnemonic="RustBindgen",
+        progress_message="Generating bindings for {}..".format(header.path),
+        env={
+            "RUST_BACKTRACE": "1",
+            # Bindgen loads libclang at runtime, so we setup LD_LIBRARY_PATH
+            "LD_LIBRARY_PATH": ":".join([f.dirname for f in clang_libs]),
+            "CLANG_PATH": clang.path,
+        },
+        executable=bindgen,
+        arguments=[args],
+        tools=[bindgen, clang],
+    )
+
+    if rustfmt:
+        ctx.actions.run_shell(
+            inputs=depset([rustfmt, unformatted]),
+            outputs=[output],
+            command="{} --write-mode=plain {} > {}".format(rustfmt.path, unformatted.path, output.path),
+            tools=[rustfmt],
+        )
+
+rust_bindgen = rule(
+    _rust_bindgen,
+    attrs = {
+        "header": attr.label(allow_single_file = True),
+        "cc_lib": attr.label(),
+        "bindgen": attr.label(
+            executable = True,
+            cfg = "host",
+            default = Label("@cargo//:cargo_bin_bindgen"),
```

Comment:
I don't think this is a good default. It isn't clear where a new user would get this binary, and "@cargo//" isn't an established convention.

Reply:
We should establish a convention, @cargo seems like the simplest. 

Should we omit the default entirely?

Reply:
I support omitting the default. 

I'm receptive to declaring a default workflow around external deps though.

Reply:
Doesn't _really_ work with libraries, might work with binaries as @$binary//:$binary, @cargo//:$binary, or @rust//:$binary..  Happy to punt on this, hasn't come up in other scenarios and only affects one parameter here.

Reply:
This is preempted by using a toolchain without defaults instead.